kmir/src/kmir/kmir.py

```diff
@@ -66,70 +67,94 @@ def kcfg_explore(self, label: str | None = None) -> Iterator[KCFGExplore]:
+            parsed_item = parser.parse_mir_json(item, 'MonoItem')
+            if not parsed_item:
+                raise ValueError(f'Could not parse MonoItemKind: {parsed_item}')
+            parsed_item_kinner, _ = parsed_item
+            if type(parsed_item_kinner) is KApply and parsed_item_kinner.label.name == 'monoItemWrapper':
+                parsed_items[item_ty] = parsed_item_kinner.args[1]
+        return map_of(parsed_items)
+
+    def _make_type_and_adt_maps(self, smir_info: SMIRInfo) -> tuple[KInner, KInner]:
+        parser = Parser(self.definition)
+        types: dict[KInner, KInner] = {}
+        adts: dict[KInner, KInner] = {}
+        for type in smir_info._smir['types']:
+            parse_result = parser.parse_mir_json(type, 'TypeMapping')
+            assert parse_result is not None
+            type_mapping, _ = parse_result
+            assert isinstance(type_mapping, KApply) and len(type_mapping.args) == 2
+            ty, tyinfo = type_mapping.args
+            if ty in types:
+                raise ValueError(f'Key collision in type map: {ty}')
+            types[ty] = tyinfo
+            if isinstance(tyinfo, KApply):
+                if tyinfo.label.name in ['TypeInfo::EnumType', 'TypeInfo::StructType']:
+                    adts[tyinfo.args[1]] = ty
+        return (map_of(types), map_of(adts))
 
     def make_call_config(
-        self, parsed_smir: KApply, smir_info: SMIRInfo, start_symbol: str = 'main', sort: str = 'GeneratedTopCell'
+        self, smir_info: SMIRInfo, start_symbol: str = 'main', sort: str = 'GeneratedTopCell', init: bool = False
     ) -> tuple[KInner, list[KInner]]:
-
         if not start_symbol in smir_info.function_tys:
             raise KeyError(f'{start_symbol} not found in program')
 
-        _sym, _allocs, functions, items, types, _ = parsed_smir.args
-
         args_info = smir_info.function_arguments[start_symbol]
-
         locals, constraints = symbolic_locals(smir_info, args_info)
+        types, adts = self._make_type_and_adt_maps(smir_info)
 
-        subst = {
+        _subst = {
             'K_CELL': mk_call_terminator(smir_info.function_tys[start_symbol], len(args_info)),
             'STARTSYMBOL_CELL': KApply('symbol(_)_LIB_Symbol_String', (token(start_symbol),)),
             'STACK_CELL': list_empty(),  # FIXME see #560, problems matching a symbolic stack
             'LOCALS_CELL': list_of(locals),
-            'FUNCTIONS_CELL': KApply('mkFunctionMap', (functions, items)),
-            'TYPES_CELL': KApply('mkTypeMap', (map_empty(), types)),
-            'ADTTOTY_CELL': KApply('mkAdtMap', (map_empty(), types)),
+            'FUNCTIONS_CELL': self._make_function_map(smir_info),
+            'TYPES_CELL': types,
+            'ADTTOTY_CELL': adts,
         }
 
-        config = self.definition.empty_config(KSort(sort))
+        _init_subst: dict[str, KInner] = {}
+        if init:
+            _subst['LOCALS_CELL'] = list_empty()
```

Comment:
If `init` is set we cannot execute functions that have arguments, maybe we can add `assert len(args_info) == 0`.

Reply:
Chasing this thread down turned into quite a bit more of a rabbit hole than I'm ready to do on this PR. For now, I'm going to merge as-is. I don't think it's quite correct here, because of how calling convention works, but tests seem to be passing OK.